ecall_t.c

```diff
@@ -0,0 +1,429 @@
+{
+    n->plain_int = 100;
+    n->array_of_int = (int*)malloc(4 * sizeof(int));
+    for (int i = 0; i < 4; i++)
+        n->array_of_int[i] = i;
+    n->shallow_struct = NULL;
+    n->array_of_struct = (CountStruct*)malloc(3 * sizeof(CountSizeStruct));
+    for (int i = 0; i < 3; i++)
+    {
+        n->array_of_struct[i].ptr = (uint64_t*)malloc(3 * sizeof(uint64_t));
+        n->array_of_struct[i].count = 100;
+        n->array_of_struct[i].size = 200;
+        for (int j = 0; j < 3; j++)
+            n->array_of_struct[i].ptr[j] = data[j];
+    }
+}
+ */
+
+/* Generated structs */
+typedef struct ShallowStruct
+{
+    uint64_t count;
+    size_t size;
+    uint64_t* ptr;
+} ShallowStruct;
+
+typedef struct CountStruct
+{
+    uint64_t count;
+    size_t size;
+    uint64_t* ptr;
+} CountStruct;
+
+typedef struct NestedStruct
+{
+    int plain_int;
+    int* array_of_int;
+    ShallowStruct* shallow_struct;
+    CountStruct* array_of_struct;
+} NestedStruct;
+
+/* EDL ECALL stub */
+static void ecall_deepcopy_nested_out(
+    uint8_t* input_buffer,
+    size_t input_buffer_size,
+    uint8_t* output_buffer,
+    size_t output_buffer_size,
+    size_t* output_bytes_written)
+{
+    oe_result_t _result = OE_FAILURE;
+
+    /* Prepare parameters. */
+    deepcopy_nested_out_args_t* pargs_in =
+        (deepcopy_nested_out_args_t*)input_buffer;
+    deepcopy_nested_out_args_t* pargs_out =
+        (deepcopy_nested_out_args_t*)output_buffer;
+
+    size_t input_buffer_offset = 0;
+    size_t output_buffer_offset = 0;
+    OE_ADD_SIZE(input_buffer_offset, sizeof(*pargs_in));
+    OE_ADD_SIZE(output_buffer_offset, sizeof(*pargs_out));
+
+    /* Make sure input and output buffers lie within the enclave. */
+    /* oe_is_within_enclave explicitly checks if buffers are null or not. */
+    if (!oe_is_within_enclave(input_buffer, input_buffer_size))
+        goto done;
+
+    if (!oe_is_within_enclave(output_buffer, output_buffer_size))
+        goto done;
+
+    /* Set in and in-out pointers. */
+    /* There were no in nor in-out parameters. */
+
+    /* Set out and in-out pointers. */
+    /* In-out parameters are copied to output buffer. */
+    if (pargs_in->n)
+        OE_SET_OUT_POINTER(
+            n, ((size_t)1 * sizeof(NestedStruct)), NestedStruct*);
+
+    /* Check that in/in-out strings are null terminated. */
+    /* There were no in nor in-out string parameters. */
+
+    /* lfence after checks. */
+    oe_lfence();
+
+    /* Call user function. */
+    deepcopy_nested_out(pargs_in->n);
+
+    if (pargs_out->n)
+    {
+        NestedStruct* _rhs = pargs_in->n;
+        for (size_t _i_1 = 0; _i_1 < 1; _i_1++)
+        {
+            pargs_out->n[_i_1].plain_int = _rhs[_i_1].plain_int;
+            pargs_out->n[_i_1].shallow_struct = _rhs[_i_1].shallow_struct;
+            pargs_out->n[_i_1].array_of_int = NULL;
+            if (_rhs[_i_1].array_of_int)
+            {
+                int* _l_1_array_of_int =
```

Comment:
Own comment: use local variable to prevent the host from manipulating the memory used by the `memcpy` below. Note that the `oe_host_malloc` ensures that the return buffer resides on the host.